pkg/sbom/scanner/scanner.go

```diff
@@ -86,129 +121,187 @@ func (s *Scanner) enoughDiskSpace(opts sbom.ScanOptions) error {
+// On shutdown, it also shutdowns the scanQueue.
+func (s *Scanner) startCacheCleaner(ctx context.Context) {
+	cleanTicker := time.NewTicker(config.Datadog.GetDuration("sbom.cache.clean_interval"))
+	defer func() {
+		cleanTicker.Stop()
+		s.running = false
+	}()
 	go func() {
-		cleanTicker := time.NewTicker(config.Datadog.GetDuration("sbom.cache.clean_interval"))
-		defer cleanTicker.Stop()
-		s.running = true
-		defer func() { s.running = false }()
-	loop:
 		for {
 			select {
-			// We don't want to keep scanning if image channel is not empty but context is expired
 			case <-ctx.Done():
-				break loop
+				s.scanQueue.ShutDown()
+				return
 			case <-cleanTicker.C:
+				s.cacheMutex.Lock()
+				log.Debug("cleaning SBOM cache")
 				for _, collector := range collectors.Collectors {
 					if err := collector.CleanCache(); err != nil {
 						_ = log.Warnf("could not clean SBOM cache: %v", err)
 					}
 				}
-			case request, ok := <-s.scanQueue:
-				// Channel has been closed we should exit
-				if !ok {
-					break loop
-				}
-				telemetry.SBOMAttempts.Inc(request.Collector(), request.Type())
-
-				collector := request.collector
-				if err := s.enoughDiskSpace(request.collector.Options()); err != nil {
-					var imgMeta *workloadmeta.ContainerImageMetadata
-					if store := workloadmeta.GetGlobalStore(); store != nil {
-						img, err := store.GetImage(request.ID())
-						if err != nil {
-							imgMeta = img
-						}
-					}
-
-					result := sbom.ScanResult{
-						ImgMeta: imgMeta,
-						Error:   fmt.Errorf("failed to check current disk usage: %w", err),
-					}
-					request.sendResult(&result)
-					telemetry.SBOMFailures.Inc(request.Collector(), request.Type(), "disk_space")
-					continue
-				}
+				s.cacheMutex.Unlock()
+			}
+		}
+	}()
+}
 
-				scanTimeout := request.collector.Options().Timeout
-				if scanTimeout == 0 {
-					scanTimeout = defaultScanTimeout
-				}
+func (s *Scanner) start(ctx context.Context) {
+	if s.running {
+		return
+	}
+	s.running = true
+	s.startCacheCleaner(ctx)
+	s.startScanRequestHandler(ctx)
+}
 
-				scanContext, cancel := context.WithTimeout(ctx, scanTimeout)
-				createdAt := time.Now()
-				scanResult := collector.Scan(scanContext, request.ScanRequest)
-				generationDuration := time.Since(createdAt)
-				scanResult.CreatedAt = createdAt
-				scanResult.Duration = generationDuration
-				if scanResult.Error != nil {
-					telemetry.SBOMFailures.Inc(request.Collector(), request.Type(), "scan")
-				} else {
-					telemetry.SBOMGenerationDuration.Observe(generationDuration.Seconds(), request.Collector(), request.Type())
-				}
-				cancel()
-				request.sendResult(&scanResult)
-				if request.collector.Options().WaitAfter != 0 {
-					t := time.NewTimer(request.collector.Options().WaitAfter)
-					select {
-					case <-ctx.Done():
-					case <-t.C:
-					}
-					t.Stop()
-				}
+func (s *Scanner) startScanRequestHandler(ctx context.Context) {
```

Comment:
Should we check for ctx.Done() in this function? Like we do in `startCacheCleaner` for example.

Reply:
I am not sure it's possible as `scanQueue.Get()` is a blocking call. I think we can only call `scanQueue.Shutdown()` and we do it in the cache cleaner on shutdown.

Reply:
I see. I think that's fine.
I would probably expect to see the shutdown in `startScanRequestHandler` instead of the cleaner, but it's just personal preference :+1:

Reply:
I think so as well. Perhaps I can move it to a 3rd goroutine